controllers/bookController.js

```diff
@@ -12,14 +17,46 @@ exports.addBook = catchAsync(async (req, res, next) => {
   });
 });
 exports.getAllBooks = catchAsync(async (req, res, next) => {
-  const books = await Book.find();
+  const queryObj = { ...req.query };
+  const excludedFields = ["page", "sort", "limit"];
+
+  // Exclude fields that are not for filtering
+  excludedFields.forEach((el) => delete queryObj[el]);
+
+  // Advanced filtering
+  let queryStr = JSON.stringify(queryObj);
+  queryStr = queryStr.replace(/\b(gte|gt|lte|lt)\b/g, (match) => `$${match}`);
+
+  // Initialize the query with filtering
+  let query = Book.find(JSON.parse(queryStr));
+
+  // Pagination
+  const page = req.query.page * 1 || 1;
+  const limit = req.query.limit * 1 || 10;
+  const skip = (page - 1) * limit;
+  query = query.skip(skip).limit(limit);
+
+  // Execute the query
+  const books = await query;
+  const totalItems = await Book.countDocuments(JSON.parse(queryStr));
+
+  if (req.query.page) {
+    if (skip >= totalItems)
+      return next(new AppError("This page does not exist", 404));
```

Comment:
You don't have to throw error that page does not exist. Just return an empty array for books

Reply:
alright, noted